CMakeLists.txt

```diff
@@ -102,62 +101,81 @@ if (ENABLE_CC_XPLAT_TRACE_SH)
-    unset(NO_ICU_PATH_GIVEN_SH CACHE)
-    unset(ICU_INCLUDE_PATH_SH CACHE)
-    unset(ICU_INCLUDE_PATH CACHE)
+if(NO_ICU_SH)
+    set(NO_ICU 1)
+    unset(NO_ICU_SH CACHE)
 endif()
 
-function(clr_unknown_arch)
-    message(FATAL_ERROR "Only AMD64, ARM and I386 are supported")
-endfunction()
+if(SYSTEM_ICU_SH)
+    set(SYSTEM_ICU 1)
+    unset(SYSTEM_ICU_SH CACHE)
+endif()
 
-if(NOINTL_ICU_SH)
-    unset(NOINTL_ICU_SH CACHE)    # don't cache
-    set(ICU_INTL_DISABLED 1)
+if(INTL_ICU_SH)
+    unset(INTL_ICU_SH CACHE)
+    set(INTL_ICU 1)
+else()
+    unset(INTL_ICU_SH CACHE)
+    set(INTL_ICU 0)
 endif()
 
 if(ICU_INCLUDE_PATH)
     add_definitions(-DHAS_REAL_ICU=1)
-    set(ICU_CC_PATH "${ICU_INCLUDE_PATH}/../lib/")
-    find_library(ICUUC icuuc PATHS ${ICU_CC_PATH} NO_DEFAULT_PATH)
-    if(NOT ICU_INTL_DISABLED)
+    set(ICU_LIBRARY_PATH "${ICU_INCLUDE_PATH}/../lib/")
+    find_library(ICUUC icuuc PATHS ${ICU_LIBRARY_PATH} NO_DEFAULT_PATH)
+
+    if(INTL_ICU)
         add_definitions(-DINTL_ICU=1)
-        find_library(ICU18 icui18n PATHS ${ICU_CC_PATH} NO_DEFAULT_PATH)
-        # icu header files are either located under the same folder or i18n is under a relative path
-        # TODO (unlikely): shall we add `--icu-intl` to build.sh ?
+        find_library(ICUIN icui18n PATHS ${ICU_LIBRARY_PATH} NO_DEFAULT_PATH)
+        # In a default install, ICU header files are all in ICU_ROOT/include
+        # However, for Node, the include/ folder is never generated, so we have to look
+        # in NODE/deps/icu/source/{common|i18n} for headers
         set(ICU_INCLUDE_PATH
-          "${ICU_INCLUDE_PATH}"
-          "${ICU_INCLUDE_PATH}/../i18n/")
+            "${ICU_INCLUDE_PATH}"
+            "${ICU_INCLUDE_PATH}/../i18n/"
+        )
     endif()
+
     if(ICUUC)
-      message("-- found ICU libs: ${ICU_CC_PATH}")
-      find_library(ICUDATA icudata PATHS ${ICU_CC_PATH} NO_DEFAULT_PATH)
-      if (NOT ICUDATA)
-          set(ICUDATA "")
-      endif()
-      set(ICULIB
-        ${ICUUC}
-        ${ICU18}
-        ${ICUDATA}
+        message("-- found ICU libs: ${ICU_LIBRARY_PATH}")
+        find_library(ICUDATA icudata PATHS ${ICU_LIBRARY_PATH} NO_DEFAULT_PATH)
+        if (NOT ICUDATA)
+            set(ICUDATA "")
+        endif()
+        set(ICU_LIBRARIES
+            ${ICUUC}
+            ${ICUIN}
+            ${ICUDATA}
         )
     endif()
 endif()
 
+if(CC_TARGET_OS_LINUX OR CC_TARGET_OS_ANDROID)
+    if(SYSTEM_ICU OR (NOT ICU_INCLUDE_PATH AND NOT NO_ICU))
+        set(ICU_LIBRARIES "icuuc")
+        if(INTL_ICU)
+            add_definitions(-DINTL_ICU=1)
+            set(ICU_LIBRARIES
+                ${ICU_LIBRARIES}
+                icui18n
+            )
+        endif()
+        add_definitions(-DHAS_REAL_ICU=1)
+    endif()
+elseif(CC_TARGET_OS_OSX)
+    # in case ICU path was given but it doesn't exist, build script will fail.
+    # so, fallback only if ICU path wasn't given
+    if(NOT ICU_INCLUDE_PATH)
```

Comment:
Why this is OSX specific?

Reply:
It used to be as well -- it was just further down in the file. I moved it up to be together with the rest of the ICU configuration.

Reply:
With that being said, I would probably support making the behavior for garbage ICU_INCLUDE_PATHs consistent across platforms.